src/ngraph/pass/nop_elimination.cpp

```diff
@@ -134,62 +134,263 @@ static bool eliminate_reshape_v1(const std::shared_ptr<Node>& node)
     return false;
 }
 
+static size_t count_unknown_dims(const PartialShape& ps)
+{
+    size_t rc = 0;
+    if (ps.is_static())
+    {
+        return rc;
+    }
+    for (auto i = 0; i < ps.rank().get_length(); i++)
+    {
+        if (ps[i].is_dynamic())
+        {
+            rc += 1;
+        }
+    }
+    return rc;
+}
+
+static bool replace_squeeze_unsqueeze(const std::shared_ptr<Node>& node)
+{
+    auto shape_ps = node->get_output_partial_shape(0);
+    if (shape_ps.rank().get_length() == 0)
+    {
+        return false;
+    }
+    if (count_unknown_dims(shape_ps) > 1)
+    {
+        return false;
+    }
+    std::vector<int64_t> target_shape;
+    for (auto i = 0; i < shape_ps.rank().get_length(); i++)
+    {
+        if (shape_ps[i].is_dynamic())
+        {
+            target_shape.emplace_back(-1);
+        }
+        else
+        {
+            target_shape.emplace_back(shape_ps[i]);
+        }
+    }
+
+    shared_ptr<Node> reshape;
+    auto input = node->input_value(0).get_node_shared_ptr();
+    auto pat =
+        op::Constant::create<int64_t>(element::i64, Shape{target_shape.size()}, target_shape);
+
+    if (is_type<opset3::Reshape>(input) || is_type<opset3::Squeeze>(input) ||
+        is_type<opset3::Unsqueeze>(input))
+    {
+        reshape = make_shared<opset3::Reshape>(input->input_value(0), pat, false);
+    }
+    else
+    {
+        reshape = make_shared<opset3::Reshape>(node->input_value(0), pat, false);
+    }
+
+    // skip if reshape is nop
+    if (reshape->get_input_partial_shape(0).same_scheme(shape_ps))
```

Comment:
I don't like the idea that we have a nop Reshape elimination here. This transformation should not care about this since we should have a separate transformation which will remove all redundant reshapes. 

If there are no more other significant comments then please ignore this one and proceed with the merge.